common/workunit/workflow.cpp

```diff
@@ -613,12 +776,816 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+        CCloneWorkflowItem & cur = static_cast<CCloneWorkflowItem&>(workflow->queryWfid(i));
+
+        if((cur.queryNumPredecessors() == 0) && cur.isActive())
+        {
+        #ifdef TRACE_WORKFLOW
+                LOG(MCworkflow, "item %u has been added to the inital task queue", cur.queryWfid());
+        #endif
+            itemQueue.push(i);
+        }
+    }
+    itemQueueSem.signal(itemQueue.size());
+}
+bool WorkflowMachine::checkIfDone()
+{
+    if((pendingContingencies == 0) && (rootReached))
+    {
+        if( (branchCount == 0) || failureClauseOld)
+        {
+        #ifdef TRACE_WORKFLOW
+                LOG(MCworkflow, "workflow done");
+        #endif
+            done = true;
+            itemQueueSem.signal(numThreads);
+            return true;
+        }
+    }
+    return false;
+}
+void WorkflowMachine::processWfItems()
+{
+    Owned<WorkflowException> error;
+    while(!done)
+    {
+        itemQueueSem.wait();
+        if(!done)
+        {
+            unsigned currentWfid = 0;
+            //enter critical section
+            {
+                CriticalBlock thisBlock(queueCritSec);
+                currentWfid = itemQueue.front();
+                itemQueue.pop();
+            }
+            try
+            {
+                executeItemParallel(currentWfid);
+            }
+            catch(WorkflowException * e)
+            {
+                error.setown(e);
+                done = true;
+                break;
+            }
+        }
+    }
+    if(error)
+        throw error.getLink();
+}
+void WorkflowMachine::performParallel(IGlobalCodeContext *_ctx, IEclProcess *_process)
+{
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "starting perform parallel");
+    #endif
+    ctx = _ctx;
+    process = _process;
+
+    //relink workflow
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Starting to mark Items with their successors");
+    #endif
+    addSuccessors();
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Finished marking Items with their successors");
+    #endif
+
+    //add initial values
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Adding initial workflow items");
+    #endif
+    initialiseItemQueue();
+
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Initialising threads");
+    #endif
+
+    //initialise thread count
+    numThreads = getThreadNumFlag();
+    if(numThreads < 1)
+            numThreads = 4;
+    unsigned maxThreads = getAffinityCpus();
+    if(numThreads > maxThreads)
+        numThreads = maxThreads;
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "num threads = %u", numThreads);
+    #endif
+
+    std::vector<std::thread *> threads(numThreads);
+    //loop through and start threads
+    for(int i=0; i < numThreads; i++)
+        threads[i] = new std::thread([this]() {  this->processWfItems(); });
+
+
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Calling join threads");
+    #endif
+    //loop through and call join
```

Comment:
this is an example of a comment that says what the code is doing (which you can deduce from the code), rather than why.  Better would be
// wait for threads to process the workflow items, and then exit when all the work is done.